text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,345 @@
+
+```toml
+[workspace]
+members = ["relative/path/to/child1", "../child2"]
+
+# or ...
+
+[package]
+workspace = "../foo"
+```
+
+The root of a workspace, indicated by the presence of `[workspace]`, is
+responsible for defining the entire workspace (listing all members).
+This example here means that two extra crates will members of the workspace
+(which also includes the root).
+
+The `package.workspace` key is used to point at a workspace root. For
+example this Cargo.toml indicates that the Cargo.toml in `../foo` is the
+workspace root that this package is a member of.
+
+These keys are mutually exclusive when applied in `Cargo.toml`. A crate may
+*either* specify `package.workspace` or specify `[workspace]`. That is, a
+crate cannot both be a root in a workspace (contain `[workspace]`) and also be
+member of another workspace (contain `package.workspace`).
+
+### "Virtual" `Cargo.toml`
+
+A good number of projects do not necessarily have a "root `Cargo.toml`" which is
+an appropriate root for a workspace. To accommodate these projects and allow for
+the output of a workspace to be configured regardless of where crates are
+located, Cargo will now allow for "virtual manifest" files. These manifests will
+currently **only** contains the `[workspace]` table and will notably be lacking
+a `[project]` or `[package]` top level key.
+
+Cargo will for the time being disallow many commands against a virtual manifest,
+for example `cargo build` will be rejected. Arguments that take a package,
+however, such as `cargo test -p foo` will be allowed. Workspaces can eventually
+get extended with `--all` flags so in a workspace root you could execute
+`cargo build --all` to compile all crates.
+
+### Validating a workspace
+
+A workspace is valid if these two properties hold:
+
+1. A workspace has only one root crate (that with `[workspace]` in
+   `Cargo.toml`).
+2. All workspace crates defined in `workspace.members` point back to the
+   workspace root with `package.workspace`.
+
+While the restriction of one-root-per workspace may make sense, the restriction
+of crates pointing back to the root may not. If, however, this restriction were
+not in place then the set of crates in a workspace may differ depending on
+which crate it was viewed from. For example if workspace root A includes B then
+it will think B is in A's workspace. If, however, B does ont point back to A,
+then B would not think that A was in its workspace. This would in turn cause the
+set of crates in each workspace to be different, futher causing `Cargo.lock` to
+get out of sync if it were allowed. By ensuring that all crates have edges to
+each other in a workspace Cargo can prevent this situation and guarantee robust
+builds no matter where they're executed in the workspace.
+
+To alleviate misconfiguration Cargo will emit an error if the two properties
+above hold for any crate attempting to be part of a workspace. For example, if
```

Comment:
Is this supposed to say that Cargo will emit an error if the two properties *do not* hold?